ksql-api/src/main/java/io/confluent/ksql/api/server/AcksSubscriber.java

```diff
@@ -54,22 +55,41 @@ protected void afterSubscribe(final Subscription subscription) {
   }
 
   @Override
-  public void handleValue(final JsonObject value) {
+  public void handleValue(final InsertResult result) {
     checkContext();
-    insertsStreamResponseWriter.writeInsertResponse();
-    acksSent++;
-    if (insertsSent != null && insertsSent == acksSent) {
-      close();
-    } else if (response.writeQueueFull()) {
-      if (!drainHandlerSet) {
-        response.drainHandler(v -> {
-          drainHandlerSet = false;
-          checkMakeRequest();
-        });
-        drainHandlerSet = true;
+    if (responseEnded) {
+      return;
+    }
+    if (result.succeeded()) {
+      insertsStreamResponseWriter.writeInsertResponse(new InsertAck(result.sequenceNumber()));
+      acksSent++;
+      if (insertsSent != null && insertsSent == acksSent) {
+        close();
+      } else if (response.writeQueueFull()) {
+        if (!drainHandlerSet) {
+          response.drainHandler(v -> {
+            drainHandlerSet = false;
+            checkMakeRequest();
+          });
+          drainHandlerSet = true;
+        }
+      } else {
+        checkMakeRequest();
       }
     } else {
-      checkMakeRequest();
+      log.error("Error in processing inserts", result.exception());
+      final InsertError insertError;
+      final Exception exception = result.exception();
+      if (exception instanceof KsqlInsertsException) {
+        insertError = new InsertError(result.sequenceNumber(),
+            ((KsqlInsertsException) exception).getErrorCode(),
+            exception.getMessage());
+      } else {
+        insertError = new InsertError(result.sequenceNumber(), ERROR_CODE_INTERNAL_ERROR,
+            "Error in processing inserts");
+      }
+      insertsStreamResponseWriter.writeError(insertError).end();
+      responseEnded = true;
```

Comment:
I find such heavily nested if/else if/else code very hard to grok. 

Have you considered moving the two main chunks into `handleSuccess` and `handleError` type methods?

Taking the success one as an example, it might look something like:

```java
private void handleSuccess(final InsertResult result) {
      insertsStreamResponseWriter.writeInsertResponse(new InsertAck(result.sequenceNumber()));
     acksSent++;

     if (insertsSent != null && insertsSent == acksSent) {
         close();
         return;
      }

      if (response.writeQueueFull()) {
            if (!drainHandlerSet) {
                 response.drainHandler(v -> {
                     drainHandlerSet = false;
                     checkMakeRequest();
                 });
               drainHandlerSet = true;
            }
            return;
      }

     checkMakeRequest();
}
```

I find this much easier to follow, as each `if` block returns, meaning I don't need to scan down the rest of the method to see what else is going to happen.

Reply:
ack